[M2-5609] Update DashboardTable/Team Listing View

JustinNusca opened the pull request:
### 📝 Description

🔗 [M2-5609](https://mindlogger.atlassian.net/browse/M2-5609): [FE][Teams Overview] Team Listing View

This PR updates the appearance of the Dashboard Table to match the specs from Figma. While the associated task is intended for the "Team Listing View", the table UI itself is reused across many of the table views in the various applet tabs, and some of the UI changes I've made impacts those screens as well.

In addition to the changes to the Dashboard Table UI, which mostly consist of border and padding updates, I've also tweaked the columns shown on the Team Listing view. This includes:

- Removing the pins column/ability to pin a row, per the Figma specs.
- Updating the "Role" column to be titled in the singular, and to render the role as a `Chip` component.
- Adding a "Checkmark" column, intended for performing bulk operations
- Removing the title of the "Actions" column
- Re-ordering columns

### 📸 Screenshots

#### Team Listing

| Before | After |
|-|-|
| ![team-before](https://github.com/ChildMindInstitute/mindlogger-admin/assets/1670836/d237251b-bff8-405c-9676-44cf267943c7) | ![team-after](https://github.com/ChildMindInstitute/mindlogger-admin/assets/1670836/21a44196-0b37-4533-9fd9-02618a411cf1) |


[M2-5609]: https://mindlogger.atlassian.net/browse/M2-5609?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JustinNusca committed:
M2-5609: Update Chip component styles.

- This updates the appearance of the component to match specs in Figma.

JustinNusca committed:
M2-5609: Add additional palette variable.

JustinNusca committed:
M2-5609: Update `DashboardTable` styles.

JustinNusca committed:
M2-5609: Update `TableHead` variant styles.

This variant is used by the DashboardTable component, which does not have bottom borders for header rows.

JustinNusca committed:
M2-5609: Update Managers table columns.

JustinNusca committed:
M2-5609: Add `className` prop to `TableHead`.

JustinNusca committed:
M2-5609: Update `DashboardTable` on "Participants" screen.

IhorNesterenko99 commented on src/modules/Dashboard/components/DashboardTable/DashboardTable.styles.ts:
```diff
@@ -19,8 +23,11 @@ export const StyledTableCell = styled(TableCell, shouldForwardProp)`
-    return `
-      height: 4.8rem;
-    `;
+    return { paddingTop: 12, paddingBottom: 12 };
```
Please use theme.spacing for paddings and margins. Please also check other occurrences.

IhorNesterenko99 commented on src/modules/Dashboard/features/Managers/Managers.tsx:
```diff
@@ -116,52 +107,110 @@ export const Managers = () => {
-                  menuItems={getManagerActions(actions, filteredManager)}
-                  data-testid="dashboard-managers-table-actions"
-                />
+                <Box display="flex" justifyContent="flex-end" alignItems="center" width="100%">
```
Please consider using StyledComponents for styling instead of inline properties to maintain consistency with the current codebase.

farmerpaul commented on src/modules/Dashboard/components/DashboardTable/DashboardTable.tsx:
```diff
@@ -78,9 +78,13 @@ export const DashboardTable = ({
-                      : row[key].content(row)}
+                    <Box
+                      sx={{ overflow: 'hidden', textOverflow: 'ellipsis', whiteSpace: 'nowrap' }}
+                    >
```
Looks like there's a preexisting shared component for this very purpose: `StyledEllipsisText`

Also just a note that I've seen it suggested in past reviews to restrict inline styling (preferably via `sx`, as you've done here) to no more than 2 attributes; for more than that, it warrants its own Styled component.

farmerpaul commented on src/modules/Dashboard/features/Managers/Managers.utils.tsx:
```diff
@@ -17,42 +19,61 @@ export const getHeadCells = (id?: string): HeadCell[] => {
 
```
This comment is referring to `ManagersColumnsWidth` enum above. ☝🏻
It seems you've removed any references to the enum (which I can understand being unnecessary if the values are not being repeated anywhere), but as such we can get rid of the enum definition, too.

farmerpaul commented on src/modules/Dashboard/features/Managers/Managers.tsx:
```diff
@@ -116,52 +107,110 @@ export const Managers = () => {
+          <Box component="span" sx={{ color: variables.palette.outline_variant2 }}>
+            --
+          </Box>
+        );
```
Please use the existing styled component, `StyledMaybeEmpty`, for this purpose, which you'll find is used in a couple other spots (such as [Participants table](https://github.com/ChildMindInstitute/mindlogger-admin/blob/8aee5c3f90428f1bb779ff1723ec3bf6e34c057f/src/modules/Dashboard/features/Participants/Participants.tsx#L335)).

farmerpaul commented on src/modules/Dashboard/features/Managers/Managers.tsx:
```diff
@@ -116,52 +107,110 @@ export const Managers = () => {
+            setSelectedManager(user || null);
+            setEditAccessPopupVisible(true);
+          },
+        };
```
Could these be defined once at the `useMemo` function body level rather than for each element of the array?

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.const.tsx:
```diff
@@ -1,5 +1,6 @@
 export enum ParticipantsColumnsWidth {
-  Pin = '4.8rem',
+  Pin = '8rem',
+  Actions = '3.2rem',
```
This newly added `Actions` enum value doesn't seem to be referenced? There's already a `Menu` enum value for use with the actions column, though it has a different width than what you have here…

JustinNusca replied:
Holdover from a previous attempt at modifying the column widths that I neglected to remove 😅

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -276,9 +284,12 @@ export const Participants = () => {
+                height="20"
+                id="checkbox-empty-outlined"
+                width="20"
+              />
```
It seems we'll be standardizing on this styling of checkboxes at least in tables, and probably universally in the admin panel as a whole. Would it be worth creating a shared `Checkbox` component that defaults to this SVG for the `icon` prop?

farmerpaul commented on src/shared/styles/variables/palette.ts:
```diff
@@ -14,6 +14,7 @@ export const palette = {
   on_surface_variant: '#42474e',
   outline: '#72777f',
   outline_variant: '#c4c9d0',
+  outline_variant2: '#C2C7CF',
```
Nit 😇 
```suggestion
  outline_variant2: '#c2c7cf',
```

farmerpaul commented on src/shared/styles/theme.ts:
```diff
@@ -289,7 +289,8 @@ export const theme = createTheme({
           color: variables.palette.on_surface_variant,
-          padding: '0.6rem 1.2rem',
+          padding: '0.2rem 0.8rem',
+          height: 24,
```
```suggestion
          height: '2.4rem',
```

farmerpaul commented on src/shared/styles/theme.ts:
```diff
@@ -305,16 +306,8 @@ export const theme = createTheme({
-            a: {
-              color: variables.palette.on_surface_variant,
-            },
+            backgroundColor: variables.palette.inverse_on_surface,
```
Thanks for doing this styling! You may be aware already, but a more comprehensive update of the Chip component is going to be captured in [M2-6161](https://mindlogger.atlassian.net/browse/M2-6161).

Did you confirm it's okay to remove the hover state for this color variant? Just noting that the `TagsController` looks like it might use `Chip` in a more interactive manner.

If you are confident we can remove the hover state here, then it would make sense to update this other instance of `Chip` where I manually disabled hover state: [ParticipantSnippet.tsx](https://github.com/ChildMindInstitute/mindlogger-admin/blob/8aee5c3f90428f1bb779ff1723ec3bf6e34c057f/src/modules/Dashboard/components/ParticipantSnippet/ParticipantSnippet.tsx#L35-L40)

JustinNusca replied:
I spent a bunch of time on this today, and I think I've basically taken care of [M2-6161](https://mindlogger.atlassian.net/browse/M2-6161) in doing so 😅.

In the additional commit I pushed up, I:

- Re-implemented the hover styles, but am only conditionally applying them when the Chip is interactive (more specifically, when a `tabindex=0` attribute it present, which MUI automatically adds when an `onClick` or `onDelete` prop is passed to it. This should cover the usage in `TagsController`.

- Updated my revisions to the theme so that hover states appear correct for each color (I followed existing styles where I could, but no hover states are provided for the revised styles in Figma, so I wound up improvising a little bit and did my best to stay true to the new colors).

- Altered the prop interface for Chips to allow the set of MUI chip props in order to support the "variant" prop. The old "secondary" styles used a grey border, instead of a grey background. The MUI Chip's `variant` can be either `"filled"` or `"outlined"` that is intended to control this behaviour, so opted to disambiguate the existing styles and allow for `color` and `variant` to be set separately. This was my attempt at providing some backwards compatibility with these old styles — none of the new instances of the Chip in Figma appear to _use_ an `"outlined"` variant, and I'm not using it in my implementation, but if for some reason we need to revert to the previous "secondary" Chip styling, it can be opted into on an individual basis by setting that prop.

Chips look like so, now:

<img width="168" alt="Screenshot 2024-04-18 at 6 22 13 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1670836/93b76558-5847-47e3-ad5f-ca6e15113f26">

farmerpaul replied:
That's fantastic @JustinNusca, thanks for doing that! IMO you might as well assign yourself to M2-6161 as well as add it to this PR's description (so that it gets auto-linked to this PR).

I think the only other part of completing M2-6161 would be to make sure the chips under the Account Type column on the [Participants screen](https://www.figma.com/file/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?type=design&node-id=1628-123599&mode=design&t=mywOdvfdkUyr1NOM-4) is fully styled to match Figma, which essentially just means changing the "Pending Invite" chip to "✉︎ Pending" and using the appropriate color, if that's something you wouldn't mind squeezing in?

JustinNusca replied:
Oh sure! Updated:

![Screenshot 2024-04-19 at 11 16 15 AM](https://github.com/ChildMindInstitute/mindlogger-admin/assets/1670836/2f5d9110-84fe-4a54-842f-a90286c6489f)

farmerpaul commented on src/shared/components/Table/TableHead/TableHead.types.ts:
```diff
@@ -3,6 +3,7 @@ import { HeadCell, Order } from 'shared/types/table';
 import { UiType } from '../Table.types';
 
 export type TableHeadProps = {
+  className?: string;
```
I'm not seeing where you've introduced code that uses this new `className` prop. Just curious of its purpose.

JustinNusca replied:
It's used [here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1657/commits/1baa9aab8849da8cb175ba9b57c86cd18bde5c13#diff-73665bee8296711d9fe42d0b7e90eb2d25901aa022259d7acd6b7a4465c14098R30-R33), to allow the classname that the StyledComponent creates to be passed to the StyledComponent that wraps the MUI TableHead.

farmerpaul replied:
Oh interesting, didn't realize that would be needed. TIL, thx!

farmerpaul commented on src/modules/Dashboard/features/Managers/Managers.tsx:
```diff
@@ -116,52 +107,110 @@ export const Managers = () => {
+                  <ActionsMenu
+                    menuItems={getManagerActions(actions, filteredManager)}
+                    data-testid="dashboard-managers-table-actions"
+                  />
```
I added the ability to use the new color variant of the `...` button for `ActionsMenu` a few PRs ago via `buttonColor="secondary"`. Would you mind updating it here and in the Participants table? 🙏🏻

farmerpaul reviewed (commented):
Hey @JustinNusca, I just had a handful of small suggestions/questions, but great work!!

JustinNusca committed:
M2-5609: Address PR comments.

farmerpaul commented on src/modules/Dashboard/features/Managers/Managers.utils.tsx:
```diff
@@ -8,33 +8,13 @@ import { variables } from 'shared/styles';
-          }
-        />
-      ),
+      label: <Checkbox aria-label={t('checkAll')} checked={false} />,
```
Just wondering if it was intentional to omit `data-testid="dashboard-managers-checkbox-header"`?

JustinNusca replied:
Yeah, I copied it over from the existing similar column in `Participants`, and just renamed it initially.

But this isn't used for any tests yet, because no bulk operation functionality has been implemented yet. I figured I'd remove unused attributes to reduce potential confusion, and let whoever ends up responsible for adding this feature take ownership and add it if the need it (who knows if this Checkbox will exist as-is when it comes to actually building the feature)?

farmerpaul commented on src/shared/styles/utils/index.ts:
```diff
@@ -0,0 +1,60 @@
+        color: variables.palette.on_surface_variant,
+        [key]: variables.palette.inverse_on_surface,
+
+        '&[tabindex="0"]:hover': {
```
Maybe helpful to include a brief comment (maybe outside the `switch`) to explain how the tabindex trick works here (what you mentioned in the PR comment)?

JustinNusca replied:
Added a comment, [here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1657/commits/48f16b23b2d473c8eac4bc3440b286d8fdb6ddb4#diff-afff194d342e79376a7b2c5c8e2b8e8a417a62ad672a514afe6fe8e17a4a6fdaR22-R24)!

farmerpaul commented on src/shared/styles/utils/index.ts:
Nit: This is just convention, but as yet we haven't been defining functions directly in an `index.ts` in this app, and they're mostly used to export modules defined in sibling files.

I think in this case, since this function pertains directly to theming, I'd probably place this code in `src/shared/styles/theme.utils.ts` (which is an existing convention).

farmerpaul commented on src/modules/Dashboard/components/DashboardTable/DashboardTable.tsx:
```diff
@@ -78,9 +78,13 @@ export const DashboardTable = ({
-                      : row[key].content(row)}
+                    <Box
+                      sx={{ overflow: 'hidden', textOverflow: 'ellipsis', whiteSpace: 'nowrap' }}
+                    >
```
Just checking if you're still planning to use `StyledEllipsisText` here in a future commit…

JustinNusca replied:
Whoops, I think I mixed the original comment up with `StyledMaybeEmpty` while making updates and overlooked it — [updated, here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1657/commits/48f16b23b2d473c8eac4bc3440b286d8fdb6ddb4#diff-fced307897174d1f11dc037c5464650394ae7b42ef41cd611af2c63ed4bd1e8fR82-R86)

farmerpaul reviewed (approved):
Just a few last minor comments, but preapproving!

JustinNusca committed:
M2-5609: Address additional PR comments.

- Use `StyledEllipsisText`.
- Update appearance of "Pending" type in Participants table.
- Relocate theme util & add comments.
